Mute monitor logs from AccumulatingProgressMonitor.beginTask

Follow up on https://github.com/eclipse-platform/eclipse.platform.ui/pull/1115.

While running latest I Build with our application tests I observe a lot of cases where clients can't do anything meaningful to prevent error log polluted with "beginTask should only be called once per instance" warnings from AccumulatingProgressMonitor.

In the stack below framework started beginTask in the progress dialog while waiting on a job and in the job client code simply used given job monitor to call SubMonitor.convert(monitor). This is fully valid use case, but there is no way to avoid the warning logged.

```
!ENTRY org.eclipse.jface 2 0 2023-09-20 20:32:21.203
!MESSAGE beginTask should only be called once per instance. At least call done() before further invocations
!STACK 0
java.lang.IllegalStateException: beginTask should only be called once per instance. At least call done() before further invocations
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:148)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.lambda$0(ProgressManager.java:265)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.beginTask(ProgressManager.java:265)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:483)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:449)
	at some.client.method()
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.IllegalStateException: beginTask(, 1000) was called here previously
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:155)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.addProgressListener(ProgressManager.java:244)
	at org.eclipse.ui.internal.progress.ProgressMonitorFocusJobDialog$3.runInUIThread(ProgressMonitorFocusJobDialog.java:213)
	at org.eclipse.ui.progress.UIJob.lambda$0(UIJob.java:148)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5040)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4520)
```

Another case is worse. Again, client code starts to work, calls beginTask() at very beginning of a long operation and later on passes given monitor to IResource.refreshLocal() which calls SubMonitor.convert(monitor) and that calls beginTask() that triggers a warning again. Once again, it is absolutely valid to pass any given monitor to IResource.refreshLocal() and not doing this would break the cancellation of the possibly long running operation.

```
java.lang.IllegalStateException: beginTask should only be called once per instance. At least call done() before further invocations
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:148)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.lambda$0(ProgressManager.java:265)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.beginTask(ProgressManager.java:265)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:483)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:449)
	at org.eclipse.core.internal.resources.Resource.refreshLocal(Resource.java:1596)
	at some.client.code.calling.refreshLocal()
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.IllegalStateException: beginTask(, 1000) was called here previously
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:155)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.lambda$0(ProgressManager.java:265)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.beginTask(ProgressManager.java:265)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:483)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:449)
	at some.client.code.called.from.WorkspaceJob.runInWorkspace()
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
	... 2 more
```

So basically clients are unable now to pass monitors to any Resource related API without getting a warning logged.
I haven't analyzed all the warnings yet (there are literally hundreds of them), but that above alone is enough.

So I would propose to mute the logging from AccumulatingProgressMonitor by default and enable it on demand with additional VM option.

I will provide a PR.

--- a/bundles/org.eclipse.jface/src/org/eclipse/jface/operation/AccumulatingProgressMonitor.java
+++ b/bundles/org.eclipse.jface/src/org/eclipse/jface/operation/AccumulatingProgressMonitor.java
@@ -50,6 +50,8 @@ import org.eclipse.swt.widgets.Display;
 	 * The display.
 	 */
 	private Display display;
+	private static final boolean LOG_BEGIN_TASK = Boolean
+			.getBoolean("AccumulatingProgressMonitor.logBeginTaskViolations"); //$NON-NLS-1$
 
 	/**
 	 * The collector, or <code>null</code> if none.
@@ -58,7 +60,8 @@ import org.eclipse.swt.widgets.Display;
 
 	private String currentTask = ""; //$NON-NLS-1$
 
-	private volatile Exception taskStarted;
+	private volatile boolean taskStarted;
+	private volatile Exception taskStartedStack;
 
 	private class Collector implements Runnable {
 		private String taskName;
@@ -144,15 +147,20 @@ import org.eclipse.swt.widgets.Display;
 
 	@Override
 	public void beginTask(final String name, final int totalWork) {
-		if (taskStarted != null) {
-			Exception e = new IllegalStateException(
-					"beginTask should only be called once per instance. At least call done() before further invocations", //$NON-NLS-1$
-					taskStarted);
-			Policy.getLog().log(Status.warning(e.getLocalizedMessage(), e));
+		if (taskStarted) {
+			if (LOG_BEGIN_TASK) {
+				Exception e = new IllegalStateException(
+						"beginTask should only be called once per instance. At least call done() before further invocations", //$NON-NLS-1$
+						taskStartedStack);
+				Policy.getLog().log(Status.warning(e.getLocalizedMessage(), e));
+			}
 			done(); // workaround client error
 		}
-		taskStarted = new IllegalStateException(
-				"beginTask(" + name + ", " + totalWork + ") was called here previously"); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+		if (LOG_BEGIN_TASK) {
+			taskStartedStack = new IllegalStateException(
+					"beginTask(" + name + ", " + totalWork + ") was called here previously"); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+		}
+		taskStarted = true;
 		synchronized (this) {
 			collector = null;
 		}
@@ -190,13 +198,14 @@ import org.eclipse.swt.widgets.Display;
 
 	@Override
 	public void done() {
-		if (taskStarted == null) {
+		if (!taskStarted) {
 			// ignore call to done() if beginTask() was not called!
 			// Otherwise an otherwise already started delegate would be finished
 			// see https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/61
 			return;
 		}
-		taskStarted = null;
+		taskStarted = false;
+		taskStartedStack = null;
 		synchronized (this) {
 			collector = null;
 		}

